src/content/post/fastrace-a-modern-approach-to-distributed-tracing-in-rust.mdx

```diff
@@ -0,0 +1,374 @@
+
+### 2. Performance Impact for Libraries
+
+Library authors are particularly sensitive to performance overhead, as their code may be called in tight loops or performance-critical paths. `tokio-rs/tracing`'s overhead can be substantial when enabled, which creates a dilemma:
+
+1. Always enable tracing (and impose overhead on all users)
+2. Don't instrument at all (and lose observability)
+3. Create a complex feature-flag system (increasing maintenance burden)
+
+    This is a common pattern in libraries using `tokio-rs/tracing`:
+
+    ```rust
+    #[cfg_attr(feature = "tracing", tracing::instrument(skip(password), fields(user_id = user.id)))]
+    async fn authenticate(user: &User, password: &str) -> Result<AuthToken, AuthError> {
+        // ...more code...
+    }
+    ```
+
+With `tokio-rs/tracing`, there's no clean way to have zero-cost tracing when disabled. This makes library authors reluctant to add instrumentation to performance-sensitive code paths.
+
+### 3. No Context Propagation
+
+Distributed tracing requires propagating context across service boundaries, but `tokio-rs/tracing` leaves this largely as an exercise for the developer. For example, the [tonic official example](https://github.com/hyperium/tonic/blob/master/examples/src/tracing/server.rs) demonstrates how to trace gRPC service using tonic:
+
+```rust
+// Server setup with `tokio-rs/tracing`
+Server::builder()
+    // Basic span only, no context extraction
+    .trace_fn(|_| tracing::info_span!("grpc_server"))
+    .add_service(MyServiceServer::new(MyService::default()))
+    .serve(addr)
+    .await?;
+```
+
+The above example only creates a basic span but doesn't extract tracing context from the incoming request.
+
+## Introducing `fastrace`: A Modern Solution
+
+`fastrace` takes a different approach to tracing in Rust. Its core design principles include:
+
+### 1. Zero-cost Abstraction
+
+`fastrace` is designed as a true zero-cost abstraction. When disabled, instrumentation compiles away completely, resulting in no runtime overhead. This makes it ideal for libraries concerned about performance.
+
+### 2. Ecosystem Compatibility
+
+`fastrace` focuses exclusively on distributed tracing. Through its composable design, it integrates seamlessly with the existing Rust ecosystem, including compatibility with the standard `log` crate. This architectural approach allows libraries to implement comprehensive tracing while preserving their users' freedom to use their preferred logging setup.
+
+### 3. Simplicity First
+
+The API is designed to be intuitive and require minimal boilerplate, focusing on the most common use cases while still providing extensibility when needed.
+
+### 4. Insanely Fast
+
+![Fastrace Performance](assets/trace-100-spans.png)
+
+`fastrace` is designed for high-performance applications. It can handle massive amount of spans per second with minimal impact on CPU and memory usage.
```

Comment:
```suggestion
`fastrace` is designed for high-performance applications. It can handle massive amounts of spans with minimal impact on CPU and memory usage.
```

Here is no concrete number so "per second" is wordy.